stonesoup/sensor/actionable.py

```diff
@@ -0,0 +1,179 @@
+
+        Returns
+        -------
+        bool
+            True if timestamp has been successfully set, False otherwise.
+        """
+        raise NotImplementedError()
+
+    @property
+    def _actionable_properties(self):
+        """Dictionary of all name - property pairs where the property is an
+        :class:`~.ActionableProperty` (i.e. it is modified via action)."""
+        return {_name: _property for _name, _property in type(self).properties.items()
+                if isinstance(_property, ActionableProperty)}
+
+    def _default_action(self, name, property_, timestamp):
+        """Returns the default action of the action generator associated with the property
+        (assumes the property is an :class:`~.ActionableProperty`."""
+        generator = property_.generator_cls(owner=self,
+                                            attribute=name,
+                                            start_time=self.timestamp,
+                                            end_time=timestamp)
+        return generator.default_action
+
+    def actions(self, timestamp: datetime.datetime, start_timestamp: datetime.datetime = None
+                ) -> Set[ActionGenerator]:
+        """Method to return a set of action generators available up to a provided timestamp.
+
+        A generator is returned for each actionable property that the sensor has.
+
+        Parameters
+        ----------
+        timestamp: datetime.datetime
+            Time of action finish.
+        start_timestamp: datetime.datetime, optional
+            Time of action start.
+
+        Returns
+        -------
+        : set of :class:`~.ActionGenerator`
+            Set of action generators, that describe the bounds of each action space.
+        """
+
+        if not self.validate_timestamp():
+            self.timestamp = timestamp
+
+        if start_timestamp is None:
+            start_timestamp = self.timestamp
+
+        generators = set()
+        for name, property_ in self._actionable_properties.items():
+            generators.add(property_.generator_cls(owner=self,
+                                                   attribute=name,
+                                                   start_time=start_timestamp,
+                                                   end_time=timestamp))
+        return generators
+
+    def add_actions(self, actions: Sequence[Action]) -> bool:
+        """Add actions to the sensor
+
+        Parameters
+        ----------
+        actions: sequence of :class:`~.Action`
+            Sequence of actions that will be executed in order
+
+        Returns
+        -------
+        bool
+            Return True if actions accepted. False if rejected.
+
+        Raises
+        ------
+        NotImplementedError
+            If sensor cannot be tasked.
+
+        Notes
+        -----
+        Base class returns True
+        """
+
+        if not self.validate_timestamp():
+            return
```

Comment:
Does this mean it returns neither True or False if timestamp is invalid? If so should mention that possibility in the 'Returns' section above.